Upgrade to CUDNN RNN v8 APIs

kaixih opened the pull request:
This PR updates the code to use cudnnRNNForward/cudnnRNNBackwardxxx_v8 when the CUDNN Version >= 8.0 and CudnnRNNV3 operation is used.



fyi. @nluehr

kaixih committed:
Upgrade to CUDNN RNN v8 APIs

sanjoy commented on tensorflow/stream_executor/cuda/cuda_dnn.cc:
```diff
@@ -1683,6 +1691,28 @@ port::StatusOr<DeviceMemory<uint8>> CreateRnnWorkspace(
+  uint8* begin_ptr = reinterpret_cast<uint8*>(seq_lengths.data());
+  std::vector<uint8> seq_lengths_buf(
+      begin_ptr, begin_ptr + seq_lengths.size() * sizeof(int));
+  stream->ThenMemcpyH2D<uint8>(seq_lengths_buf, dev_seq_lengths);
```
Just so I understand, you're relying on this memcpy being synchronous (otherwise you'd somehow have to keep `seq_lengths_buf` alive until the memcpy completed)?

kaixih replied:
You're right, and this is also my concern. I think for the small seqLengthsArray, the CUDA can make sure the memcpy is doing what it is expected. However, it might be risky when it is a large array.
 Alternatively, I am thinking we can keep the seqLengthsArray on device by removing the `.HostMemory("sequence_lengths")` for `CudnnRNNV3`. And use ThenMemcpyD2H (if we have this function) to copy back to host a seqLengthsArray so as to create the input tensor descriptor.

kaixih replied:
Or, is there any method to keep two copies of sequence_lengths that one is on device and the other is on host?

kaixih replied:
Figured out a new way to place the seq_lengths array into the rnn_seq_tensor object to make sure it can live long enough during the computation.

sanjoy commented on tensorflow/stream_executor/cuda/cuda_dnn.cc:
```diff
@@ -1181,6 +1181,8 @@ class CudnnRnnDescriptor : public dnn::RnnDescriptor {
     }
 
     // Create the params handle.
+    // TODO(kaixih): Can be removed when cudnnRNNForward*** and
+    // cudnnRNNForward***Ex are deprecated.
```
I could be missing something, but why is removing something dependent on deprecating some API?  Do you mean to to say "Can be removed when alternatives for API XYZ is available."?

kaixih replied:
This is meant to say that "We can remove this when cudnnRNNForward*** and cudnnRNNForward***Ex are removed from the code since the new API doesn't need it."

kaixih commented on tensorflow/stream_executor/cuda/cuda_dnn.cc:
```diff
@@ -1764,6 +1794,82 @@ port::Status CudnnSupport::DoRnnForwardImpl(
 
   SE_RETURN_IF_ERROR(CheckRNNParameterSize(cudnn, rnn_desc, input_desc));
+
+  // We use the cudnnRNNForward only when the CUDNN Version >= 8.0 and
+  // CudnnRNNV3 operation is used. To identify CudnnRNNV3 usage, we check if the
+  // sequence_lengths parameter is given.
```
Yes, I think one major reason is that we are going to deprecate the old CUDNN RNN APIs. I will add more comments.

kaixih commented:
@sanjoy I have updated the code by placing the host seq_lengths in `CudnnRnnSequenceTensorDescriptor` to make sure it is still valid when the H2D copy happens. I also updated the comments. PTAL.

sanjoy commented on tensorflow/stream_executor/cuda/cuda_dnn.cc:
```diff
@@ -1514,9 +1517,15 @@ class CudnnRnnSequenceTensorDescriptor
         parent, max_seq_length, batch_size, data_size, data_type,
         std::move(data_desc), std::move(tensor_desc));
+#if CUDNN_VERSION >= 8000
+    // Keep a copy of seq_lengths in the object to make it live long enough
```
What guarantees that `rnn_seq_tensor` will stay alive until the H2D copy is complete?  Please add a comment to the `CopyRnnSeqLengthsH2D` function since IMO that's a natural questions folks later reading this code will have.

kaixih replied:
I don't think this is guaranteed. I think it still might be destroyed before the H2D copy occurs.

I reverted this commit and add a BlockHostUntilDone call after the H2D copy to enforce the execution.

sanjoy commented on tensorflow/stream_executor/cuda/cuda_dnn.cc:
```diff
@@ -1181,6 +1181,9 @@ class CudnnRnnDescriptor : public dnn::RnnDescriptor {
     }
 
     // Create the params handle.
+    // TODO(kaixih): Should be removed when cudnnRNNForward*** and
```
Maybe add a `TODO(kaixih@nvidia.com)`?  Otherwise by default folks would assume we're talking about an ex-Googler.

kaixih replied:
Done.

kaixih commented on tensorflow/stream_executor/cuda/cuda_dnn.cc:
```diff
@@ -1539,6 +1558,7 @@ class CudnnRnnSequenceTensorDescriptor
   TensorDescriptor handle_;
   RNNDataDescriptor rnn_data_handle_;
   std::vector<cudnnTensorDescriptor_t> handles_;  // Copies of handle_.
+  std::vector<uint8> seq_lengths_array_;
```
The reason why we use uint8 rather int is because on the device side, we reuse the workspace allocator which only supports AllocateBytes and gives us DeviceMemory<uint8>. So, to call the ThenMemcpyH2D<uint8>(), we need to make sure the src and the dst have the same dtype.

I added some comment to explain this before the cast.

kaixih commented:
After a second thought, I think we might be able to remove the dependence on the `BlockHostUtilDone()` in the current PR. Since the `sequence_lengths` is passed in via `.HostMemory("sequence_lengths")`, it could be guaranteed it will stay long enough for the H2D copy.

The current implement (before the PR) is to copy it into input_descriptor (which is a local pageable location). And this PR's strategy is to fetche it from the input_descriptor and copy it to device, which requires the sync op.

I think we can do something different by passing the sequence_lengths to the compute method so that we can call H2D copy over it directly, which will require no sync anymore.

I will push a commit later today.

kaixih commented:
@sanjoy I have refactored the code. PTAL.

sanjoy commented on tensorflow/core/kernels/cudnn_rnn_ops.cc:
```diff
@@ -826,14 +826,33 @@ Status DoForward(OpKernelContext* context, const RnnDescriptor& rnn_desc,
+
+  Tensor seq_lengths_tensor;
+  DeviceMemory<int> seq_lengths_ptr;
+  if (sequence_lengths != nullptr) {
+	  auto seq_lengths_vec = sequence_lengths->template flat<int>();
```
Can we use the `absl::Span<const int> seq_lengths` created above?

kaixih replied:
Yes, this is a good point. I just removed them in the new commit. PTAL.

sanjoy commented on tensorflow/core/kernels/cudnn_rnn_ops.cc:
```diff
@@ -905,17 +924,35 @@ Status DoBackward(
+  Tensor seq_lengths_tensor;
+  DeviceMemory<int> seq_lengths_ptr;
+  if (sequence_lengths != nullptr) {
+	  auto seq_lengths_vec = sequence_lengths->template flat<int>();
```
Same here: can we use the `absl::Span<const int> seq_lengths` created above?

sanjoy reviewed (changes requested):
LGTM modulo the small nits mentioned inline.

kaixih committed:
Cleanup codes
